### ISSUE TITLE
Positioning of client mappers

### ISSUE THREAD

**Issue by Captain-P-Goldfish** [keycloak/keycloak-ui#3342]
### Description

we just had some troubles finding the client mappers within the new keycloak admin console. We managed to find them but its pretty not really obvious. May it be possible to place the mappers back in the outliner of the client next to the client scopes item?

**Comment by jbman** [comment]
It would be really good to have tab "Mappers" and there all mappers listed which are applied, including their source client scope where they are managed. This would give a nice overview of all the mappers which are applied for the client.

The tab would look like this, but as a client tab
![image](https://user-images.githubusercontent.com/789496/190985150-b224bf5b-6fa2-47e3-a179-00e98a9887b6.png)

The "...-dedicated" scope would also occur here under "Managed at client scope".

**Comment by xianli123** [comment]
Thanks @jonkoops @jbman @Captain-P-Goldfish , I took a look at this issue. I remember this design is a compromise. In the future, the Scopes and Mappers may be removed from clients. We could involve @stianst to discuss it. What do you think? @stianst

**Comment by Captain-P-Goldfish** [comment]
what reason could there be to remove the scopes and mappers from the clients? This is actually a pretty powerful feature and removing it would be rather problematic for some of our usecases.

**Comment by xianli123** [comment]
Hi @Captain-P-Goldfish We plan to remove mappers and scope (inside client), and only allow assigning client scopes to a client. As for more details, I will convey your concerns to the developer team and get back to you as soon as possible.

**Comment by Captain-P-Goldfish** [comment]
thx. Would be great if the mappers would be kept. We are using them  the mappers for two purposes: 
1. to add hardcoded details into the access-token to clients that authenticate with client_credentials grant. These hardcoded details are necessary for some additional operations within the authenticating application.
2. To map additional audiences into the access-token

**Comment by jbman** [comment]
Hi @xianli123 and @Captain-P-Goldfish, 

I think it would be ok to remove dedicated client mapper and role scope configuration, so that a mappers and roles scope are always configured as an entity which can be used at multiple clients. 

However the view for one client should still support to check the complete configuration of a client by keeping the "Mapper" and "Scope" tabs at the client. The mapper tab should show an aggregated view of all mappers managed at client scope entities of the client. This is what I suggest in my comment above. This view allows an admin to check which data could end up in the tokens for this client. It also allows for navigating to the client scope to add or remove a mapper there.

btw.: I suggest to name the tab "Scope" either "Role Scope" or "Client Role Scope" to make it more distinguishable from the client scope entity.

**Comment by Captain-P-Goldfish** [comment]
> I think it would be ok to remove dedicated client mapper and role scope configuration, so that a mappers and roles scope are always configured as an entity which can be used at multiple clients.

what about data that is not supposed to be added to multiple clients?
Using client scopes is possible but it would harm the clarity of the configuration if different scopes are created for different clients. In my opinion it would be better to keep the possibility to add explicit meta-datasets into the access tokens of specific clients that are only meant for this client only. 

And what about audience-mappers? What is gonna replace them?

**Comment by jbman** [comment]
That's true, a client scope can be named myClient-client-scope but it doesn't prevent that the scope is used by other clients. But isn't a naming convention enough to signalize that the scope isn't intended to be shared?
Is anything special about the audience mapper?

I think it's good to have the client-scope configuration sharable by default, as you want to be able to hand out a second client which gets the same token claims as the first one (e.g. for secret rotation, or for testing a new client scope without modification of the existing client). But this isn't my main point.I'm also totally fine with keeping the client dedicated client scope. 

My main point is full visibility of all configured mappers for the admin. Admin shouldn't need to go to each client-scope one by one to compose the list of mappers which produce token claims.

**Comment by xianli123** [comment]
@Captain-P-Goldfish @jbman Thanks for your ideas and suggestions. I have conveyed your comments to the developer. If you have any other comments, please feel free to leave them here.

**Comment by stianst** [comment]
@mposolda FIY

**Comment by stianst** [comment]
We are planning on deprecating/removing the ability to add mappers and scope directly to a client, and rather make it only possible to add these via a client scope. It just doesn't make much sense to have two separate implementations for adding mappers and scope to clients.

At the UI level how to represent it is a bit unclear. I could see that we could still want some form of a "dedicated" or single client client scope of sorts that stands out more than the rest. I can also see what @jbman is suggesting with an easy way to see the "effective" mappers and scope for a given client easier.

**Comment by ssilvert** [comment]
@xianli123 Will take a closer look to figure out what we should do.

**Comment by xianli123** [comment]
Hi @jbman @Captain-P-Goldfish @stianst @ssilvert, let me elaborate on the requirements and conclusions we have reached. 

1. We should be able to see all the mappers and scopes that are effective for a particular client. 
2. In addition, client-side dedicated mappers and scopes should stand out from other mappers for now. 
3. In the future, we won't allow users to add mappers and scopes directly to the client, although we haven't done so yet.

Do these points include all you want to achieve? Or are there any points that you want to change or add? Please let me know. If we reach the conclusions, I will update the design based on the latest requirements. Thanks.

**Comment by Captain-P-Goldfish** [comment]
Hi @xianli123,

this sounds perfectly sufficient to me.

**Comment by xianli123** [comment]
@Captain-P-Goldfish Thanks for your confirmation. I will update the design asap.

**Comment by jbman** [comment]
@xianli123 Sounds good!
Would just add the navigation requirement to point 1: 
We should be able to see all the mappers and scopes that are effective for a particular client and navigate to the configuration of mapper or scope.

**Comment by xianli123** [comment]
Hi @stianst, what do you think about these functionalities below? Can we implement them?

>1. We should be able to see all the mappers and scopes that are effective for a particular client and navigate to the configuration of mapper or scope.

> 2. In addition, client-side dedicated mappers and scopes should stand out from other mappers and scopes for now.

> 3. In the future, we won't allow users to add mappers and scopes directly to the client, although we haven't done so yet.

**Comment by xianli123** [comment]
Hi @jbman， I found the effective mappers and scops can be viewed in the **Evaluate** tab. Have you checked it before?
<img width="1077" alt="Screenshot 2023-02-03 at 19 31 53" src="https://user-images.githubusercontent.com/53931519/216593616-ea3f7690-de60-40dd-9974-a396dcd74dd8.png">

**Comment by stianst** [comment]
This area can certainly be improved on, and even if we stop adding mappers and scope directly to a client it should still be easy to add mappers/scope for a single client. That could for instance be through a dedicate client scope for the client for example, which is something that can be enabled/disabled easily. 

It's probably not something we'll get to anytime soon though, but perhaps there's a way we can make the mapper/scope for a given client stand out more, which is still compatible with the idea of in the future having a dedicated client scope for a given client?

**Comment by jbman** [comment]
Hi @xianli123, the Evaluate feature shows effective mappers and the Client scope they are defined within. The navigation part is missing. But the main disadvantage is that you have to add all scope parameters manually to get the fulll set of amppers which may be active. 
![image](https://user-images.githubusercontent.com/789496/216614094-079863b3-4e2a-4f42-952a-a92b4ddd63ff.png)

So as workaround this might be ok, but hard to learn. A view which has the full effective set by default would be much better. 

Additional Note: I already argued in the design phase, that "Evaluate" itself would be much better as top level client tab. It allows to view the resulting token even if you don*'t have learned anything about how client scopes and mappers are responsible for token content. has also the benefit that these second level tabs "Setup" and "Evaluate" aren't needed at all.

**Comment by xianli123** [comment]
Hi @jbman, sorry for the late reply. As for the effective mappers and scopes, I can optimize the design. 

With regard to the "Evaluate", I am not sure if we can change its' priority. What do you think? @stianst

**Comment by jbman** [comment]
@xianli123 Do you mean an optimized design for the Client Mappers tab? This is the one used for managing clients and I still would prefer  **full visibility of all configured mappers for the admin** at this tab. Here its' not about effective mappers for a specific requested scope. It is about showing the full set which could be effective,

Moving the evaluate tab one level up and/or using it to get all effective mappers may really be a different issue.

**Comment by xianli123** [comment]
@jbman I agree with you that the evaluate tab should be a different issue. 

As for mappers and scopes, I mean we can show all the effective mappers in a separate place. As is shown below, all effective mappers of the client are listed in a tab. The labels in front of the mappers indicate the original client scopes. Does it make sense?
<img width="817" alt="Screenshot 2023-02-09 at 13 19 56" src="https://user-images.githubusercontent.com/53931519/217724850-b84f153f-eda9-42c5-95f6-2e8a5c104cfb.png">

**Comment by mposolda** [comment]
@xianli123 @jbman +1 for moving `Evaluate` tab as top level tab of the client.

However when we talk about "effective mappers", I think it is important to also always consider the `scope` parameter as "effective mappers" used in the token are different based on it's value. My reason for that:

- Let's assume that tab `effective mappers` will automatically show all mappers (even those, which can be added by "optional client scopes" and hence specifically requested by `scope` parameter). This may cause confusions IMO as admin will see all "effective mappers", which won't be shown in the token by default. Typical example is `phone` or `address` as those are "optional client scopes" by default. Hence admin can be confused why the `phone` is not shown in his token by default (EG. "Why is phone claim not shown in my token even if it is shown in effective mappers???")

- Let's assume the opposite: the tab `effective mappers` won't automatically show optional mappers. This is better option, but still not ideal as admin won't know that he can request `phone` claim by adding `scope=phone` to the OIDC request

In shortcut: "effective mappers" and "effective role scopes" makes sense to me always just with the combination with `scope` parameter.

**Comment by xianli123** [comment]
Thanks, @mposolda. As for "Evaluate", I can rethink it. And we may need to create a new issue.

When it comes to "Effective mappers", does determining whether a mapper is **effective** depends on whether the user has selected the client scope that contains the mapper? In this case, is it not so meaningful to show all mappers?

**Comment by xianli123** [comment]
Thanks, @mposolda. As for "Evaluate", I can rethink it. And we may need to create a new issue.

When it comes to the "effective mappers" and "effective scopes", could you please elaborate on the application scenarios?

According to your explanation, let me share my understanding of the usage of them (not 100% sure). 
1. If the admin wants to generate ID tokens or access tokens for a specific user, the admin needs to select the `scope parameters` in the **Evaluate** tab and specifies a user.
2. The generated ID token or access token will include the info on effective mappers and scopes.
3. The effective mappers and scopes depend on what `scope parameters` are selected.

Are all the things I listed above right or accurate?

**Comment by mposolda** [comment]
@xianli123 The admin currently needs to select the user if he wants to simulate generate ID token (or access token) for that user. For seeing the "effective mappers", the admin currently does not need to select the user.

My point is, that during real authentication, if you want to see `phone` claim in the ID token (which means that you want `phone number` mapper to be "effective mapper"), you need to use `scope=phone` and explicitly attach it to the request. The `evaluate` tab simulates how the token would look like for the particular user and what are "effective mappers" used during generation of that token.

IMO it would be misleading to display "phone number" by default in the "effective mappers" without showing that you actually need to use `scope=phone` in the request. That's what I've tried to describe in the previous comment.

So I don't think that "effective mappers" should show by default list of all mappers from both "default" and "optional" client scopes without showing which values of "scope parameter" values are needed to generate those.

In shortcut, my vote would be to keep the content of `Evaluate` tab as is and not remove the `Scope parameter` field from it.

**Comment by xianli123** [comment]
@mposolda Thanks. I understood.
